pkg/nodeagent/controller/operatingsystemconfig/containerd_config.go

```diff
@@ -0,0 +1,289 @@
+					}
+				}
+
+				return append(imports, importPath), nil
+			},
+		},
+		{
+			name: "sandbox image",
+			path: containerdConfigPaths[containerdConfigFileVersion][sandboxImagePath],
+			setFn: func(value any) (any, error) {
+				if criConfig.Containerd == nil {
+					return value, nil
+				}
+
+				return criConfig.Containerd.SandboxImage, nil
+			},
+		},
+		{
+			name: "CNI plugin dir",
+			path: containerdConfigPaths[containerdConfigFileVersion][cniPluginPath],
+			setFn: func(_ any) (any, error) {
+				return cniPluginDir, nil
+			},
+		},
+	}
+
+	if criConfig.CgroupDriver != nil {
+		patches = append(patches, patch{
+			name: "cgroup driver",
+			path: containerdConfigPaths[containerdConfigFileVersion][cgroupDriverPath],
+			setFn: func(_ any) (any, error) {
+				return *criConfig.CgroupDriver == extensionsv1alpha1.CgroupDriverSystemd, nil
+			},
+		})
+	}
+
+	if criConfig.Containerd != nil {
+		for _, pluginConfig := range criConfig.Containerd.Plugins {
+			patches = append(patches, patch{
+				name: "plugin configuration",
+				path: replacePluginPath(append(structuredmap.Path{"plugins"}, pluginConfig.Path...), pluginPathReplacements, containerdConfigFileVersion),
+				setFn: func(val any) (any, error) {
+					switch op := ptr.Deref(pluginConfig.Op, extensionsv1alpha1.AddPluginPathOperation); op {
+					case extensionsv1alpha1.AddPluginPathOperation:
+						values, ok := val.(map[string]any)
+						if !ok || values == nil {
+							values = map[string]any{}
+						}
+
+						pluginValues := pluginConfig.Values
+						// Return unchanged values if plugin values is not set, i.e. only create table.
+						if pluginValues == nil {
+							return values, nil
+						}
+
+						if err := json.Unmarshal(pluginValues.Raw, &values); err != nil {
+							return nil, err
+						}
+
+						return values, nil
+					case extensionsv1alpha1.RemovePluginPathOperation:
+						// Return nil if operation is remove, to delete the entire sub-tree.
+						return nil, nil
+					default:
+						return nil, fmt.Errorf("operation %q is not supported", op)
+					}
+				},
+			})
+		}
+	}
+
+	for _, p := range patches {
+		if err := structuredmap.SetMapEntry(content, p.path, p.setFn); err != nil {
+			return fmt.Errorf("unable setting %q in containerd config.toml: %w", p.name, err)
+		}
+	}
+
+	f, err := r.FS.OpenFile(configFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("unable to open containerd config.toml: %w", err)
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Error(err, "Failed closing file", "file", f.Name())
+		}
+	}()
+
+	return toml.NewEncoder(f).Encode(content)
+}
+
+func isConfigPathPrefix(path, prefix structuredmap.Path) bool {
+	if len(prefix) > len(path) {
+		return false
+	}
+
+	return slices.Equal(prefix, path[:len(prefix)])
+}
+
+func replaceConfigPathPrefix(path, prefix, replace structuredmap.Path) structuredmap.Path {
+	if slices.Equal(replace, prefix) {
+		return path
+	}
```

Comment:
I do not understand the need for this check. It is a bit irritating to read and it should be handled fine in the rest of the function.
If the path includes the prefix it can be replaced without any issues and if it does not include the prefix the next `if` condition will have an early return anyway.
In summary, I would prefer to remove this check for better readability.
WDYT?